client/tso_service_discovery.go

```diff
@@ -526,3 +503,84 @@ func (c *tsoServiceDiscovery) findGroupByKeyspaceID(
 
 	return resp.KeyspaceGroup, nil
 }
+
+func (c *tsoServiceDiscovery) getTSOServer(sd ServiceDiscovery) (string, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	var (
+		addrs []string
+		err   error
+	)
+	t := c.tsoServerDiscovery
+	if len(t.addrs) == 0 || t.failureCount == len(t.addrs) {
+		addrs, err = sd.DiscoverMicroservice(tsoService)
+		if err != nil {
+			return "", err
+		}
+		failpoint.Inject("serverReturnsNoTSOAddrs", func() {
+			log.Info("[failpoint] injected error: server returns no tso addrs")
+			addrs = nil
+		})
+		if len(addrs) == 0 {
+			// There is no error but no tso server address found, which means
+			// either the server side is experiencing some problems to get the
+			// tso primary addresses or the server hasn't been upgraded to the
+			// version which processes and returns GetClusterInfoResponse.TsoUrls.
+			// In this case, we fall back to the old way of discovering the tso
+			// primary addresses from etcd directly.
+			log.Warn("[tso] no tso server address found,"+
```

Comment:
This file permission has changed to 0755, need to change it back.

Reply:
changed back.